src/main/scala/org/constellation/storage/RateLimiting.scala

```diff
@@ -0,0 +1,64 @@
+package org.constellation.storage
+
+import cats.effect.concurrent.Ref
+import cats.effect.Sync
+import cats.implicits._
+import io.chrisdavenport.log4cats.Logger
+import org.constellation.primitives.{Schema, Transaction}
+
+import scala.math.ceil
+
+class RateLimiting[F[_]: Sync: Logger]() {
+  private[storage] val counter: Ref[F, Map[Schema.Address, Int]] = Ref.unsafe(Map())
+  private[storage] val blacklisted: StorageService[F, Int] = new StorageService()
+
+  def update(txs: List[Transaction]): F[Unit] =
+    for {
+      grouped <- txs.groupBy(_.src).mapValues(_.size).pure[F]
+      _ <- counter.update(_ |+| grouped)
+
+      _ <- counter.get.flatTap(c => Logger[F].debug(s"Update [rate-limiting]: $c"))
+      _ <- blacklist()
+    } yield ()
+
+  def reset(cbHashes: List[String])(checkpointService: CheckpointService[F]): F[Unit] =
+    for {
+      cbs <- cbHashes.traverse(checkpointService.fullData).map(_.flatten)
+      txs = cbs.flatMap(_.checkpointBlock.get.transactions.toList)
+      grouped = txs.groupBy(_.src).mapValues(_.size)
+      _ <- counter.set(grouped)
+
+      _ <- counter.get.flatTap(c => Logger[F].debug(s"Reset [rate-limiting]: $c"))
+      _ <- counter.get
+        .flatMap(_.map(a => available(a._1)).toList.sequence)
+        .flatTap(c => Logger[F].info(s"Available [rate-limiting]: $c"))
+      _ <- blacklist()
+    } yield ()
+
+  def available(address: Schema.Address): F[Int] =
+    for {
+      c <- counter.get
+      accounts = c.size
+      address <- counter.get.map(_.getOrElse(address, 0))
+      available = (if (accounts > 0) limits.available / accounts else limits.available) - address
+    } yield available
+
+  private def blacklist(): F[Unit] =
+    for {
+      c <- counter.get
+      left <- c.keys.toList.traverse(a => available(a).map(a -> _))
```

Comment:
`available(_)` checks mutable ref thus we can end up by not blacklisting someone who should be (on concurrent update) thus going above given threshold. I don't know if this limit is soft or we should better check this inside `modify` or `update`

Reply:
I'll add single lock in next PR